pkg/agent/cniserver/interface_configuration_windows.go

```diff
@@ -204,14 +205,27 @@ func (ic *ifConfigurator) removeContainerLink(containerID, epName string) error
 // removeHNSEndpoint removes the HNSEndpoint from HNS and local cache.
 func (ic *ifConfigurator) removeHNSEndpoint(endpoint *hcsshim.HNSEndpoint, containerID string) error {
 	epName := endpoint.Name
+	deleteCh := make(chan error)
 	// Remove HNSEndpoint.
-	_, err := endpoint.Delete()
-	if err != nil {
-		if !strings.Contains(err.Error(), notFoundHNSEndpoint) {
-			klog.Errorf("Failed to delete container interface %s: %v", containerID, err)
-			return err
+	go func() {
+		_, err := endpoint.Delete()
+		deleteCh <- err
+	}()
+
+	// Deleting HNS Endpoint is blocking in some corner cases. It might be a bug in Windows HNS service. To avoid
+	// hanging in cniserver, add timeout control in HNSEndpoint deletion.
+	select {
+	case err := <-deleteCh:
+		if err != nil {
+			if !strings.Contains(err.Error(), notFoundHNSEndpoint) {
+				klog.Errorf("Failed to delete container interface %s: %v", containerID, err)
+				return err
+			}
 		}
+	case <-time.After(1 * time.Second):
+		return fmt.Errorf("timeout when deleting HNSEndpoint %s", epName)
```

Comment:
What will happen if timeout? Will the CNI DEL request be sent again to retry deleting the HNS endpoint?

Reply:
kubelet might ignore the error in CNI Del, so there is no retry for HNS Endpoint deletion. As for HNSEndpoint, once it is timeout for a deletion, there might be a stuck in HNS service. It is an issue in Windows HNS service, so there is no need to retry delete the HNS Endpoint, as it should be timeout again.

Reply:
Ok.

Reply:
Looks there is something wrong in my understanding. I have rechecked kubelet code, kubelet might retry if it receives any error from CNI plugin. But for HNS Endpoint, if the deletion is stuck, retry of the deletion is much possibly hung again.